### DISCUSSION TITLE
Types for $fetch payload/body

### DISCUSSION THREAD

**Discussion by angelhdzmultimedia** [nuxt/nuxt#21881] [Questions] [answered]
Hi, Is there a way to type the body of a $fetch or useFetch hooks?

Or my only option is to use tRPC or GraphQL? 

Thanks.

**Comment by warflash** [comment]
You can pass generic types to them. e.g. `useFetch<MyType>(...)`

**Reply by angelhdzmultimedia** [reply]
Thank you for replying.

That's the first thing I tried before creating this thread.

![image](https://github.com/nuxt/nuxt/assets/10191085/69c80580-7e75-4913-b300-aaa4b7f086b6)

Apparently that type is for the response data, not the payload/body.

**Comment by TheAlexLichter** [comment, marked as answer]
Types for parameters/body for requests to nitro api endpoints are work in progress.

See https://github.com/unjs/h3/pull/417/ for more

**Reply by angelhdzmultimedia** [reply]
Highly appreciated! 🔥🤝🤯

**Reply by gkTim** [reply]
Since this PR is closed, should it be possible to get correct types in $fetch? I have tried it and I get typing for the response body and the api path (which is already awesome, nice work). But my body still has no typing. So maybe I simply do it wrong.

**Reply by TheAlexLichter** [reply]
@gkTim Response body should have a type "by default", yes, given the URL you provided is static and points to a nitro endpoint ☺️

**Reply by gkTim** [reply]
@manniL Thanks for the quick answer. Sry I meant the request body not the response body.

**Reply by angelhdzmultimedia** [reply]
Thanks for clarifying. I was a bit confused hehe. 😅

Response body is already type safe. Only the request body is a work in progress. 🔥👏☕

**Reply by TheAlexLichter** [reply]
Request body works as well since h3 1.8 
https://unjs.io/blog/2023-08-15-h3-towards-the-edge-of-the-web

**Reply by angelhdzmultimedia** [reply]
https://github.com/nuxt/nuxt/issues/23009#issue-1881478762

**Reply by chrisjansky** [reply]
Sorry for resurrecting this, I am a bit thick in the head: can someone please provide an example typing $fetch for both request body and response?

**Reply by TheAlexLichter** [reply]
@chrisjansky not possible yet, follow https://github.com/nuxt/nuxt/issues/23009#issue-1881478762
